security.go

```diff
@@ -0,0 +1,186 @@
+
+	"golang.org/x/net/context"
+)
+
+// SecurityMethod extracts the security method (OAuth2Security,
+// BasicAuthSecurity, APIKeySecurity or OtherSecurity) from the
+// request context.
+//
+// This is to be used from within a security middleware
+// implementation. If you do get into your middleware code, it is
+// guaranteed that the request will hold a valid SecurityMethod.  You
+// can then typecast it and validate the proper credentials.
+func SecurityMethod(ctx context.Context) interface{} {
+	return ctx.Value(securityMethodKey)
+}
+
+// Scopes extracts from a request the scopes relevant to the action
+// being executed.
+//
+// Call this from within your Security middleware implementation.
+//
+// Scopes can be empty
+func Scopes(ctx context.Context) (out []string) {
+	scopes, ok := ctx.Value(securityScopesKey).([]string)
+	if !ok {
+		return
+	}
+	return scopes
+}
+
+// securityMiddleware represents a security method middleware, used
+// for deduplicating the different security method's methods!
+type securityMiddleware struct {
+	middleware Middleware
+	method     interface{}
+
+	// MethodName is the name specified for this authentication method
+	// which will appear in the Swagger file.
+	MethodName string
+
+	// Description of the security method
+	Description string
+
+	// Metadata is some data passed on from the DSL.
+	Metadata map[string][]string
+}
+
+// Dispatch returns a wrapped Handler, configured to handle a certain
+// action's credentials validation.
+//
+// It is called by `app`-generated code. You shouldn't need to use
+// this directly.
+func (sec *securityMiddleware) Dispatch(h Handler, scopes ...string) Handler {
+	return func(ctx context.Context, rw http.ResponseWriter, req *http.Request) error {
+		if sec.middleware == nil {
+			RequestService(ctx).Error("security method not implemented", "method", sec.MethodName)
+			return errors.New("security method not implemented")
+		}
+
+		ctx = context.WithValue(ctx, securityMethodKey, sec.method)
+		if len(scopes) != 0 {
+			ctx = context.WithValue(ctx, securityScopesKey, scopes)
+		}
+
+		return sec.middleware(h)(ctx, rw, req)
+	}
+}
+
+///////////////////////////////////////////////////////////////////
+
+// OAuth2Security represents the `oauth2` security method. It is
+// automatically instantiated in your generated code when you use the
+// different `*Security()` DSL functions and `Security()` in your
+// design.
+type OAuth2Security struct {
+	securityMiddleware
+
+	// Flow defines the OAuth2 flow type. See http://swagger.io/specification/#securitySchemeObject
+	Flow string
+	// TokenURL defines the OAuth2 tokenUrl.  See http://swagger.io/specification/#securitySchemeObject
+	TokenURL string
+	// AuthenticationURL defines the OAuth2 authenticationUrl.  See http://swagger.io/specification/#securitySchemeObject
+	AuthenticationURL string
+	// Scopes defines a list of scopes for the security method.
+	Scopes []string
+}
+
+// Type returns the security method type.
```

Comment:
I meant embedding `goa.Middleware`. OK I hadn't realized there was no Kind - leave it the way it is if that ends up being too complicated.